module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+;; Guile) and one external (that is returned by hash-table-hash-function).
+;; Internal hash functions accepts two arguments, while external functions
+;; accept one argument (and possibly more optional arguments).
+
+;;; Code:
+
+;;;; Module definition & exports
+
+(define-module (ice-9 generic-hash-tables)
+  #:use-module (srfi srfi-1)	;alist-cons,second&c,assoc
+  #:use-module (srfi srfi-8)    ;receive
+  #:use-module (srfi srfi-9)
+  #:use-module (srfi srfi-9 gnu)
+  #:use-module (ice-9 format)
+  #:export (;; Type constructors & predicate
+	    make-hash-table
+            hash-table? hash-table hash-table-unfold alist->hash-table
+            ;; Reflective queries
+            hash-table-equivalence-function hash-table-hash-function
+            hash-table-weakness hash-table-key-weakness hash-table-value-weakness
+            ;; Predicates
+            hash-table-mutable? hash-table-contains?
+            hash-table-empty? hash-table=?
+            ;; Accessors
+            hash-table-ref hash-table-ref/default
+            ;; Mutators
+            hash-table-set! hash-table-set-single!
+            hash-table-delete! hash-table-delete-single!
+            hash-table-intern! hash-table-intern!/default
+            hash-table-update! hash-table-update!/default
+            hash-table-pop! hash-table-clear!
+            ;; The whole hash table
+            hash-table-size hash-table-find hash-table-count
+            hash-table-keys hash-table-values hash-table-entries
+            hash-table-key-vector hash-table-value-vector hash-table-entry-vectors
+            ;; Mapping and folding
+            hash-table-map hash-table-for-each hash-table-map! hash-table-map->list
+            hash-table-fold hash-table-prune!
+            ;; Copying
+            hash-table-copy hash-table-empty-copy
+            ;; Conversion
+            hash-table->alist
+            ;; Hash tables as sets
+            hash-table-union! hash-table-intersection! hash-table-difference!
+            hash-table-xor!
+            ;; Hashing
+            string-ci-hash hash-by-identity hash-by-value hash)
+  #:re-export (string-hash)
+  #:replace (hash make-hash-table hash-table?))
+
+
+;;;; Internal helper macros
+
+;; Define these first, so the compiler will pick them up.
+
+;; I am a macro only for efficiency, to avoid varargs/apply.
+(define-macro (hashx-invoke hashx-proc ht-var . args)
+  "Invoke HASHX-PROC, a `hashx-*' procedure taking a hash-function,
+assoc-function, and the hash-table as first args."
+  `(,hashx-proc (ht-hash-function ,ht-var)
+		(ht-associator ,ht-var)
+		(ht-real-table ,ht-var)
+		. ,args))
+
+(define-macro (with-hashx-values bindings ht-var . body-forms)
+  "Bind BINDINGS to the hash-function, associator, and real-table of
+HT-VAR, while evaluating BODY-FORMS."
+  `(let ((,(third bindings) (ht-real-table ,ht-var))
+         (,(first bindings) (ht-hash-function ,ht-var))
+	 (,(second bindings) (ht-associator ,ht-var)))
+     . ,body-forms))
```

Comment:
I would have done this with a procedure and `call-with-values`.

Reply:
> I would have done this with a procedure and `call-with-values`.

`;; I am a macro only for efficiency, to avoid varargs/apply.`